File: test/controllers/NotesTest.java
package controllers;

import com.google.common.collect.ImmutableMap;
import models.Note;
import org.junit.Before;
import org.junit.Test;
import play.mvc.Http.Response;
import play.test.Fixtures;
import play.test.FunctionalTest;

import java.util.List;

import static org.hamcrest.Matchers.*;

/**
 * Date: 17.01.2012 at 21:37
 *
 * @author Marcin Swierczynski
 */
public class NotesTest extends FunctionalTest {

	@Before
	public void setUp() throws Exception {
		Fixtures.deleteDatabase();
	}

	@Test
	public void notesListResponds() throws Exception {
		Response response = GET("/notes/list");
		assertIsOk(response);
	}

	@SuppressWarnings("unchecked")
	@Test
	public void showNotesList() throws Exception {
		Note note1 = new Note("Note1 title\n note1 content").save();
		Note note2 = new Note("Note2 title\n note2 content").save();

		Response response = GET("/notes/list");

		List<Note> notes = (List<Note>) renderArgs("notes");
		assertThat(notes.get(0), is(note1));
		assertThat(notes.get(1), is(note2));

		String responseContent = response.out.toString("UTF-8");
		assertThat(responseContent, containsString(note1.getTitle()));
		assertThat(responseContent, containsString(note1.getContentWithoutTitle()));
		assertThat(responseContent, containsString(note2.getTitle()));
		assertThat(responseContent, containsString(note2.getContentWithoutTitle()));
	}

	@Test
	public void saveNewNote() throws Exception {
		POST("/notes/saveNote", ImmutableMap.of("note.content", "Note content"));

		List<Note> allNotes = Note.findAll();
		assertThat(allNotes.size(), is(1));
		assertThat(allNotes.get(0).content, is(equalTo("Note content")));
	}

	@Test
	public void rejectNoteWithoutContent() throws Exception {
		POST("/notes/saveNote", ImmutableMap.of("note.content", ""));
		assertThat(Note.count(), is(0L));
	}

	@Test
	public void rejectNoteWithTooShortContent() throws Exception {
		POST("/notes/saveNote", ImmutableMap.of("note.content", "test"));
		assertThat(Note.count(), is(0L));
	}
}

File: test/controllers/NotesTest.scala
package controllers

import org.specs2.mutable.Specification
import play.api.test._
import play.api.test.Helpers._
import models.Note
import play.mvc.Http.Status.{BAD_REQUEST, OK, SEE_OTHER}

/**
 * Date: 22.03.2012 at 21:03
 *
 * @author Marcin Swierczynski
 */

class NotesTest extends Specification {

	"respond to the Notes action" in {
		running(FakeApplication()) {
			val result = controllers.Notes.notes(FakeRequest())
			status(result) must equalTo(OK)
		}
	}

	"show notes list" in {
		running(FakeApplication()) {
			Note.create("Note1 title\\n note1 content")
			Note.create("Note2 title\\n note2 content")

			val result = controllers.Notes.notes(FakeRequest())
			contentAsString(result) must contain("Note1 title")
			contentAsString(result) must contain("Note2 title")
			contentAsString(result) must contain("note1 content")
			contentAsString(result) must contain("note2 content")
		}
	}

	"save new note" should {
		"persist note to database" in {
			running(FakeApplication()) {
				controllers.Notes.newNote(FakeRequest().withFormUrlEncodedBody(("content", "Note1")))
				val notes: List[Note] = Note.findByContent("Note1")
				notes.size must beEqualTo(1)
				notes.head.content must beEqualTo("Note1")
			}
		}
		"redirect to notes list" in {
			running(FakeApplication()) {
				val result = controllers.Notes.newNote(FakeRequest().withFormUrlEncodedBody(("content", "Note1")))
				status(result) must equalTo(SEE_OTHER)
			}
		}
	}

	"impossible to save empty note" in {
		running(FakeApplication()) {
			val result = controllers.Notes.newNote(FakeRequest())
			status(result) must equalTo(BAD_REQUEST)
		}
	}

	"impossible to save note shorter than 5 characters" in {
		running(FakeApplication()) {
			val result = controllers.Notes.newNote(FakeRequest().withFormUrlEncodedBody(("content", "Note")))
			status(result) must equalTo(BAD_REQUEST)
		}
	}

	"delete note" in {
		running(FakeApplication()) {
			controllers.Notes.newNote(FakeRequest().withFormUrlEncodedBody(("content", "Note1")))
			val newNote: Note = Note.findByContent("Note1").head

			val result = controllers.Notes.deleteNote(newNote.id)(FakeRequest())
			status(result) must equalTo(SEE_OTHER)
			Note.findByContent("Note1").size must beEqualTo(0)
		}
	}

}
